build-system/tasks/markdown-toc/index.js

```diff
@@ -0,0 +1,189 @@
+-->`;
+
+/**
+ * @param {string} content
+ * @return {?string}
+ */
+function getFrontmatter(content) {
+  const end = content.search(/(\n-|\n#)/m);
+  if (end < 0) {
+    return null;
+  }
+  const withWhitespace = content.substr(0, end);
+  if (withWhitespace.trim().length < 1) {
+    return null;
+  }
+  return withWhitespace;
+}
+
+/**
+ * @param {string} maybeComment
+ * @return {?Object}
+ */
+function isolateCommentJson(maybeComment) {
+  if (!maybeComment) {
+    return null;
+  }
+
+  const json = maybeComment.replace('<!--', '').replace('-->', '').trim();
+  if (!json) {
+    return null;
+  }
+
+  try {
+    const options = JSON.parse(json);
+    return options;
+  } catch (_) {
+    return null;
+  }
+}
+
+/**
+ * @param {string} content
+ * @return {string}
+ */
+async function overrideToc(content) {
+  const indexOfHeader = content.indexOf(header);
+
+  if (indexOfHeader < 0) {
+    return null;
+  }
+
+  const headerEnd = indexOfHeader + header.length;
+
+  const afterHeader = content.substr(headerEnd);
+
+  const frontmatter = getFrontmatter(afterHeader);
+  const frontmatterOptions = isolateCommentJson(frontmatter);
+
+  const afterFrontmatter = frontmatter
+    ? afterHeader.substr(afterHeader.indexOf(frontmatter) + frontmatter.length)
+    : afterHeader;
+
+  const listEnd = afterFrontmatter.search(/\n[^-\s]/m);
+
+  if (listEnd < 0) {
+    return null;
+  }
+
+  // https://github.com/jonschlinkert/markdown-toc#options
+  const options = {
+    firsth1: false,
+    ...frontmatterOptions,
+    // Don't allow to override `bullets` because it breaks assumptions.
+    bullets: '-',
+  };
+
+  const overridden = [
+    content.substr(0, headerEnd),
+    frontmatter,
+    toc(content, options).content,
+    afterFrontmatter.substr(listEnd),
+  ].join('\n');
+
+  return prettier.format(overridden, {
+    ...(await prettier.resolveConfig('_.md')),
+    parser: 'markdown',
+  });
+}
+
+/**
+ * @param {string} cwd
+ * @return {Object<string, ?string>}
+ */
+async function overrideTocGlob(cwd) {
+  const glob = [
+    '**/*.md',
+    '!**/{node_modules,build,dist,dist.3p,dist.tools,.karma-cache}/**',
+  ];
+  const files = globby.sync(glob, {cwd}).map((file) => path.join(cwd, file));
+  const filesIncludingString = getStdout(
+    [`grep -irl "${task}"`, ...files].join(' ')
+  )
+    .trim()
+    .split('\n');
+
+  const result = {};
+
+  for (const filename of filesIncludingString) {
+    const content = await readFile(filename, 'utf-8');
+    const tentative = await overrideToc(content);
+
+    if (!tentative) {
+      continue;
+    }
+
+    result[filename] = tentative === content ? null : tentative;
+  }
+
+  return result;
+}
+
+async function markdownToc() {
+  const result = await overrideTocGlob('.');
+  let errored = false;
+  for (const filename in result) {
+    if (filename.indexOf('markdown-toc/') >= 0) {
+      continue;
+    }
+    const tentative = result[filename];
+    if (!tentative) {
+      log(gray(filename));
```

Comment:
nit: I suggest following this style convention, which gives you concise same-line logging as a progress indicator during local development, and less noisy logs during CI:

https://github.com/ampproject/amphtml/blob/2acbd7db88491ca43b49364de4610f235a09f26b/build-system/tasks/lint.js#L74

```suggestion
      logOnSameLineLocalDev(green('Checked: ') + filename);
```